File: 백준/Silver/2792. 보석 상자/보석 상자.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	private static int N, M;
	private static int[] arr;
	private static int max;
	private static int answer;
	
	public static void main(String[] args) throws IOException {
		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		
		arr = new int[M];
		for (int i = 0; i < M; i++) {
			int ele = Integer.parseInt(br.readLine()); 
			arr[i] = ele;
			max = Math.max(max, ele);
		}
		
		binarySearch(1, max);
		
		System.out.println(answer);
	} // end of main

	private static void binarySearch(int start, int end) {
		while(start<=end) {
			int mid = (start+end)/2;
			int possibleChild = 0;
			for (int i = 0; i < M; i++) {
				possibleChild += (arr[i] % mid == 0 ? arr[i]/mid : arr[i]/mid+1);
			}
			if(possibleChild<=N) {
				answer = mid;
				end = mid-1;
			} else {
				start = mid+1;
			}
		}
	}
} // end of class

File: 백준/Silver/2792. 보석 상자/보석 상자.py
"""
질투심이 최소가 되게 나누어주는 방식은 모두 동일하게 나누기

만약 나누다가 보석이 남으면 그 수는 불가능한것 보석수 올리기

만약 남지 않으면 일단 저장하고 보석수 낮추기

보석 수 대신 나누어 두어야 하는 사람 수로 변경

나누어 주어야 하는 사람 수가 원래 아이들 수보다 많으면 보석 늘리기

적으면 일단 저장하고 보석 줄이기
"""

import sys

input = lambda: sys.stdin.readline().strip()

N, M = map(int, input().split())
lst = []
for _ in range(M):
    lst.append(int(input()))

answer = max(lst)
start = 1
end = max(lst)
while start <= end:
    mid = (start + end) // 2
    cnt_child = 0
    for gem in lst:
        if gem % mid == 0:
            cnt_child += gem // mid
        else:
            cnt_child += gem // mid + 1
    if cnt_child > N:
        start = mid + 1
    else:
        answer = min(answer, mid)
        end = mid - 1

print(answer)